alphagov/pay-publicapi
# Pp 4769 receive payment notifications

This PR adds an endpoint to publicapi (v1/payment_notifications) to handle telephone payments.

The JSON payload is represented by the CreateTelephonePayment class. It looks like this:

`{
  "amount": 12000,
  "reference": "MRPC12345",
  "description": "New passport application",
  "created_date": "2018-02-21T16:04:25Z",
  "authorised_date": "2018-02-21T16:05:33Z",
  "processor_id": "183f2j8923j8",
  "provider_id": "17498-8412u9-1273891239",
  "auth_code": "666",
  "payment_outcome": {
    "status": "success",
    "code": "P0010",
    "supplemental": {
      "error_code": "ECKOH01234",
      "error_message": "textual message describing error code"
    }
  },
  "card_type": "master-card",
  "name_on_card": "Jane Doe",
  "email_address": "jane_doe@example.com",
  "card_expiry": "02/19",
  "last_four_digits": "1234",
  "first_six_digits": "654321",
  "telephone_number": "+447700900796"
}`

When the JSON payload is successfully validated it returns **only** a 201 response. When validation is unsuccessful it returns a 422 with a response body provided by PaymentError. The required fields are:

- Amount
- Reference
- Description
- Processor id
- Provider id
- Card Type
- Card Expiry
- Last Four Digits
- First Six Digits

Amount has both minimum and maximum validation matching that of CreateCardPaymentRequest. Both reference and description have a maximum character length of 255. Both Processor id and provider id have a @NotNull annotation. (There is an edge case where the user submits the wrong type - we will have to use a custom Jackson deserializer to handle this specifically in a later story).

There are custom validators defined for Card Type, Last Four Digits and First Six Digits. (Note for card expiry I haven't enforced the requirement that it must be later than today's date).

## How to test

I have written a representation test for the deserialization of the JSON payload in CreateTelephonePaymentRequestTest. I have also written integration tests for checking that the custom validation is working correctly - see CreateTelephonePaymentCardExpiryValidation, CreateTelephonePaymentFirstSixCardDigitsValidation and CreateTelephonePaymentLastFourCardDigitsValidation in it/validation/telephone. I have also added an integration test for returning 201 - see GetTelephonePaymentIT in it/telephone.
